### PR TITLE
Add unnecessary_trailing_comma lint

### PR DESCRIPTION
**Pull request by nyurik**
This matches `cargo fmt` behavior but for `format!(...)` -style macro calls.

To make this PR easier to review, it moves all linting of the clippy code to rust-lang/rust-clippy#16533

Suggest removing an unnecessary trailing comma before the closing parenthesis in single-line format-like macro invocations (e.g. println!, format!, write!). The lint currently only runs on format-like macros because it relies on format-argument parsing; arbitrary user macros are not supported to avoid incorrect suggestions.

- Lint is in the `style` group (allow-by-default)
- Single-line only: multi-line macro invocations are not linted
- Machine-applicable fix: removes the trailing comma

```rust
// catches this
debug!("Foo={}", 1,);

// suggests this
debug!("Foo={}", 1);
```

Closes rust-lang/rust-clippy#13965

changelog: [`unnecessary_trailing_comma`]: new lint to remove trailing commas in a single-line format macro usage

### CODE REVIEW

File: clippy_lints/src/format_args.rs [hunk 1]
```diff
@@ -302,6 +332,31 @@ struct FormatArgsExpr<'a, 'tcx> {
 }
 
 impl<'tcx> FormatArgsExpr<'_, 'tcx> {
+    /// Check if there is a comma after the last format macro arg.
+    #[allow(clippy::result_large_err, reason = "due to span_to_source()")]
+    fn check_trailing_comma(&self) {
+        let sm = self.cx.sess().source_map();
+        let span = self.macro_call.span.source_callsite();
+        if !sm.is_multiline(span)
+            && let span = sm.span_extend_to_prev_char_before(span.shrink_to_hi(), ')', false)
```
Comment (notriddle) [inline clippy_lints/src/format_args.rs]: This means `println!["hi",]`, [which is allowed](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=7583cc4da04d8cfe57675b102223ffba), won't lint.

Comment (nyurik) [reply clippy_lints/src/format_args.rs]: Thx, great catch, totally forgot!!

Comment (nyurik) [reply clippy_lints/src/format_args.rs]: thx @notriddle , fixed

Comment (notriddle) [reply clippy_lints/src/format_args.rs]: This still doesn't work right in all cases. This test case gives an incorrect "correction":

```rust
    println!["Foo(,)", ]; //~ unnecessary_trailing_comma
```

**Comment by samueltardieu** [comment]
The `style` category is warn-by-default. There is a risk that this lint is too opinionated for a warn-by-default category. Also, we don't usually lint for formatting as this is handled by `rustfmt`.

**Comment by nyurik** [comment]
@samueltardieu this lint is essentially the missing piece for a single line macro calls that `cargo fmt` already does for regular function calls. I don't think anyone would ever want to have `println!("...",)` -- most of these are simply accidental leftovers for when multiline formats get reformatted as a single line.  The only reason `cargo fmt` cannot do it is that it requires type resolution to ensure correct detection. Note that `--fix` should work perfectly for this lint.  I honestly think this is a "style" as anything could possibly get.

File: clippy_lints/src/format_args.rs (hunk 1 above)
Comment (nyurik) [reply clippy_lints/src/format_args.rs]: @notriddle thx for catching all the edge cases! It took me a bit longer than anticipated. Seem to work now with all the new test cases

**Comment by Jarcho** [comment]
FCP started: https://rust-lang.zulipchat.com/#narrow/channel/257328-clippy/topic/FCP.3A.20.60unnecessary_trailing_comma.60/with/572863765

File: clippy_lints/src/format_args.rs [hunk 2]
```diff
@@ -302,6 +332,52 @@ struct FormatArgsExpr<'a, 'tcx> {
 }
 
 impl<'tcx> FormatArgsExpr<'_, 'tcx> {
+    /// Check if there is a comma after the last format macro arg.
+    #[allow(clippy::result_large_err, reason = "due to span_to_source()")]
+    fn check_trailing_comma(&self) {
+        let sm = self.cx.sess().source_map();
+        let span = self.macro_call.span.source_callsite();
```
Comment (Jarcho) [inline clippy_lints/src/format_args.rs]: Why are you calling `source_callsite` here? `macro_call.span` is already the call site of the macro.

File: clippy_lints/src/format_args.rs [hunk 3]
```diff
@@ -302,6 +332,52 @@ struct FormatArgsExpr<'a, 'tcx> {
...
+                            return Ok(span
+                                .with_lo(span.lo() + BytePos::from_usize(index + c.len_utf8()))
+                                .with_hi(span.lo() + BytePos::from_usize(last_char_index)));
+                        } else {
+                            break;
+                        }
+                    }
+                }
+                Err(SpanSnippetError::IllFormedSpan(span))
+            })
```
Comment (Jarcho) [inline clippy_lints/src/format_args.rs]: This whole thing could be:
```rust
let span_data = span.data();
if let Some(src) = span_data.get_source_text(cx)
  && let Some(src) = src.strip_suffix([')', ']', '}'])
  && let Some(src) = src.trim_end_matches(|c| c.is_whitespace() && c != '\n').strip_suffix(',')
{
    let removal_sp = SpanData {
        lo: span_data.lo + src.len(),
        hi: span_data.hi - 1,
        ..span_data
    }.span();
}
```

This will also remove the comma on weird formatting like:
```
format!("{}",
    arg,);
```

File: clippy_lints/src/format_args.rs [hunk 4]
```diff
@@ -302,6 +332,52 @@ struct FormatArgsExpr<'a, 'tcx> {
...
+        {
+            let name = self.cx.tcx.item_name(self.macro_call.def_id);
+            span_lint_and_sugg(
+                self.cx,
+                UNNECESSARY_TRAILING_COMMA,
+                removal_span,
+                format!("unnecessary trailing comma in `{name}!` macro"),
```
Comment (Jarcho) [inline clippy_lints/src/format_args.rs]: Calling out the macro name doesn't seem like adds anything useful to the message.

File: clippy_lints/src/format_args.rs [hunk 5]
```diff
@@ -229,13 +229,42 @@ declare_clippy_lint! {
     "formatting a pointer"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Suggests removing an unnecessary trailing comma before the closing parenthesis in
+    /// single-line macro invocations.
+    ///
+    /// ### Why is this bad?
+    /// The trailing comma is redundant and removing it keeps the code cleaner.
```
Comment (Jarcho) [inline clippy_lints/src/format_args.rs]: I wouldn't call this cleaner, but more consistent with regular function calls.

File: clippy_lints/src/format_args.rs (hunk 2 above)
Comment (nyurik) [reply clippy_lints/src/format_args.rs]: thx, fixed

File: clippy_lints/src/format_args.rs (hunk 3 above)
Comment (nyurik) [reply clippy_lints/src/format_args.rs]: thank you!!!  I spent over an hour trying to figure out how to optimize it, and didn't notice that strip_suffix supports a pattern!

File: clippy_lints/src/format_args.rs (hunk 4 above)
Comment (nyurik) [reply clippy_lints/src/format_args.rs]: done

**Review by nyurik — COMMENTED** [review]
thx, fixed

@rustbot ready.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/16530/0f4f81df13a18212e1b3b578ccfc8d055ad120be..ddde8fec87aa1b52debecaef800015ed7c8600d5)*

File: clippy_lints/src/format_args.rs [hunk 6]
```diff
@@ -302,6 +333,28 @@ struct FormatArgsExpr<'a, 'tcx> {
...
 impl<'tcx> FormatArgsExpr<'_, 'tcx> {
+    /// Check if there is a comma after the last format macro arg.
+    fn check_trailing_comma(&self) {
+        let span = self.macro_call.span;
+        if let Some(src) = span.get_source_text(self.cx)
+            && let Some(src) = src.strip_suffix([')', ']', '}'])
+            && let src = src.trim_end_matches(|c: char| c.is_whitespace() && c != '\n')
+            && let Some(src) = src.strip_suffix(',')
+        {
+            let src = src.trim_end_matches(|c: char| c.is_whitespace() && c != '\n');
```
Comment (Jarcho) [inline clippy_lints/src/format_args.rs]: This should either take all whitespace or not lint if this hits the start of the line.

```
print!("{}"
    , arg
    ,);
```

Either the parenthesis should be placed right after arg, or the lint shouldn't trigger. I'm fine either way, but not linting is probably better.

Comment (nyurik) [reply clippy_lints/src/format_args.rs]: @Jarcho thx, I think it is better to only handle single-line cases.

@rustbot ready

Comment (Jarcho) [reply clippy_lints/src/format_args.rs]: All this needed was 
```
&& let src = src.trim_end_matches(|c: char| c.is_whitespace() && c != '\n')
&& !src.end_with('\n')
````

This also avoids scanning the whole macro text for line breaks, which will be faster.

Comment (nyurik) [reply clippy_lints/src/format_args.rs]: thx @Jarcho , makes sense, fixed

**Comment by nyurik** [comment]
thx @Jarcho, rebased and squashed

**Review by Jarcho — APPROVED** [review]
Thank you.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/16530/e299e18e546bee90c64eba55f4276cd62633b4c0..f72e055625674d8fb2a7da4f6d484738db122ffd)*
